// @notionhq/workers/src/capabilities/oauth.ts
/**
 * Configuration for a Notion-managed OAuth provider.
 *
 * Notion owns the OAuth app credentials (client ID/secret) and the backend has
 * pre-configured provider settings.
 */
export interface NotionManagedOAuthConfiguration {
	/**
	 * The unique identifier for this OAuth provider instance.
	 */
	name: string;

	/**
	 * The pre-configured provider to use (e.g., "google", "github", "salesforce").
	 * The backend will use this to look up the OAuth configuration.
	 */
	provider: string;

	/**
	 * Optional default access token expiry (in milliseconds) to use when the OAuth provider
	 * does not return `expires_in` in token responses.
	 *
	 * Some providers (e.g. Salesforce in certain configurations) may omit expiry information.
	 */
	accessTokenExpireMs?: number;
}

/**
 * Configuration for a user-managed OAuth provider.
 *
 * You own the OAuth app credentials and must explicitly provide endpoints and
 * other OAuth parameters.
 */
export interface UserManagedOAuthConfiguration {
	/**
	 * The unique identifier for this OAuth provider instance.
	 */
	name: string;

	/**
	 * The client ID for the OAuth app.
	 */
	clientId: string;

	/**
	 * The client secret for the OAuth app.
	 */
	clientSecret: string;

	/**
	 * The OAuth 2.0 authorization endpoint URL.
	 */
	authorizationEndpoint: string;

	/**
	 * The OAuth 2.0 token endpoint URL.
	 */
	tokenEndpoint: string;

	/**
	 * The OAuth scope(s) to request.
	 */
	scope: string;

	/**
	 * Optional additional authorization parameters to include in the authorization request.
	 */
	authorizationParams?: Record<string, string>;

	/**
	 * Optional callback URL for OAuth redirect.
	 */
	callbackUrl?: string;

	/**
	 * Optional default access token expiry (in milliseconds) to use when the OAuth provider
	 * does not return `expires_in` in token responses.
	 *
	 * Some providers (e.g. Salesforce in certain configurations) may omit expiry information.
	 */
	accessTokenExpireMs?: number;
}

/**
 * Union type representing either Notion-managed or user-managed OAuth configuration.
 */
export type OAuthConfiguration =
	| NotionManagedOAuthConfiguration
	| UserManagedOAuthConfiguration;

export type OAuthCapability = ReturnType<typeof createOAuthCapability>;

/**
 * Creates an OAuth provider configuration for authenticating with third-party services.
 *
 * @param config - The OAuth configuration (Notion-managed or user-managed).
 * @returns An OAuth provider definition.
 */
export function createOAuthCapability(key: string, config: OAuthConfiguration) {
	const envKey = oauthNameToEnvKey(config.name);

	if ("provider" in config) {
		return {
			_tag: "oauth" as const,
			key,
			envKey,
			async accessToken(): Promise<string> {
				return readRequiredEnvVar(envKey, { name: config.name });
			},
			config: {
				type: "notion_managed" as const,
				name: config.name,
				provider: config.provider,
				accessTokenExpireMs: config.accessTokenExpireMs,
			},
		};
	}

	return {
		_tag: "oauth" as const,
		key,
		envKey,
		async accessToken(): Promise<string> {
			return readRequiredEnvVar(envKey, { name: config.name });
		},
		config: {
			type: "user_managed" as const,
			name: config.name,
			authorizationEndpoint: config.authorizationEndpoint,
			tokenEndpoint: config.tokenEndpoint,
			scope: config.scope,
			clientId: config.clientId,
			clientSecret: config.clientSecret,
			authorizationParams: config.authorizationParams,
			callbackUrl: config.callbackUrl,
			accessTokenExpireMs: config.accessTokenExpireMs,
		},
	};
}

function oauthNameToEnvKey(identifier: string): string {
	const encoded = Buffer.from(identifier).toString("hex").toUpperCase();

	return `OAUTH_${encoded}_ACCESS_TOKEN`;
}

function readRequiredEnvVar(key: string, context: { name: string }): string {
	const value = process.env[key];
	if (value) {
		return value;
	}

	throw new Error(
		`Missing OAuth access token env var "${key}" (name: "${context.name}"). ` +
			`Make sure you've completed OAuth for this capability and are running inside the worker runtime.`,
	);
}
