// @notionhq/workers/src/builder.ts
import type { NoticonName } from "./icon-names.js";
import type {
	DateValue,
	Icon,
	NoticonColor,
	PeopleValue,
	PlaceValue,
	RelationReference,
	TextValue,
} from "./types.js";

/**
 * Creates a rich text value.
 */
export function richText(content: string): TextValue {
	return [[content]];
}

/**
 * Creates a URL value.
 */
export function url(url: string): TextValue {
	return [[url]];
}

/**
 * Creates a title value.
 */
export function title(content: string): TextValue {
	return [[content]];
}

/**
 * Creates a text value.
 */
export function text(content: string): TextValue {
	return [[content]];
}

/**
 * Creates an email value.
 */
export function email(email: string): TextValue {
	return [[email]];
}

/**
 * Creates a phone number value.
 */
export function phoneNumber(phone: string): TextValue {
	return [[phone]];
}

/**
 * Creates a checkbox value.
 */
export function checkbox(checked: boolean): TextValue {
	return checked ? [["Yes"]] : [["No"]];
}

/**
 * Creates a file URL value.
 * @param fileUrl - The URL of the file
 * @param fileName - Optional display name for the file (defaults to URL)
 */
export function file(fileUrl: string, fileName?: string): TextValue {
	return [[fileName ?? fileUrl, [["a", fileUrl]]]];
}

/**
 * Creates a number value.
 */
export function number(value: number): TextValue {
	if (Number.isNaN(value)) {
		return [];
	}
	return [[value.toString()]];
}

/**
 * Creates a date value from a date string (YYYY-MM-DD).
 */
export function date(dateString: string): TextValue {
	validateDateString(dateString);

	const dateValue: DateValue = {
		type: "date",
		start_date: dateString,
	};

	return createDateToken(dateValue);
}

/**
 * Creates a datetime value from an ISO 8601 datetime string.
 * @param isoString - An ISO 8601 datetime string (e.g., "2024-01-15T10:30", "2024-01-15T10:30:00Z")
 * @param timeZone - Optional IANA timezone name (e.g., "America/New_York")
 */
export function dateTime(isoString: string, timeZone?: string): TextValue {
	const { date, time } = parseISODateTime(isoString);

	const dateValue: DateValue = {
		type: "datetime",
		start_date: date,
		start_time: time,
	};

	if (timeZone) {
		dateValue.time_zone = timeZone;
	}

	return createDateToken(dateValue);
}

/**
 * Creates a date range value from date strings.
 */
export function dateRange(startDate: string, endDate: string): TextValue {
	validateDateString(startDate);
	validateDateString(endDate);

	const dateValue: DateValue = {
		type: "daterange",
		start_date: startDate,
		end_date: endDate,
	};

	return createDateToken(dateValue);
}

/**
 * Creates a datetime range value from ISO 8601 datetime strings.
 * @param startDateTime - An ISO 8601 datetime string for the range start
 * @param endDateTime - An ISO 8601 datetime string for the range end
 * @param timeZone - Optional IANA timezone name (e.g., "America/New_York")
 */
export function dateTimeRange(
	startDateTime: string,
	endDateTime: string,
	timeZone?: string,
): TextValue {
	const start = parseISODateTime(startDateTime);
	const end = parseISODateTime(endDateTime);

	const dateValue: DateValue = {
		type: "datetimerange",
		start_date: start.date,
		start_time: start.time,
		end_date: end.date,
		end_time: end.time,
	};

	if (timeZone) {
		dateValue.time_zone = timeZone;
	}

	return createDateToken(dateValue);
}

/**
 * Creates a link with custom display text.
 * @param displayText - The text to display
 * @param url - The URL to link to
 */
export function link(displayText: string, url: string): TextValue {
	return [[displayText, [["a", url]]]];
}

/**
 * Creates a select value from a single option.
 */
export function select(value: string): TextValue {
	return [[value]];
}

/**
 * Creates a multi-select value from multiple options.
 * @param values - Array of option names to select
 */
export function multiSelect(...values: string[]): TextValue {
	if (values.length === 0) {
		return [];
	}
	return [[values.join(",")]];
}

/**
 * Creates a status value from a status option name.
 */
export function status(value: string): TextValue {
	return [[value]];
}

/**
 * Creates a people value from email addresses.
 * @param emails - Array of email addresses for people to include
 */
export function people(...emails: string[]): PeopleValue {
	return emails.map((email) => ({ email }));
}

/**
 * Creates a place value for a geographic location.
 * @param value - The place value with lat/lon coordinates and optional name/address
 */
export function place(value: PlaceValue): PlaceValue {
	if (typeof value.lat !== "number" || typeof value.lon !== "number") {
		throw new Error("Place value must have numeric lat and lon coordinates");
	}
	return value;
}

/**
 * Creates a relation reference from a primary key of a related record.
 * Use an array of relation references as the property value.
 *
 * @param primaryKey - The primary key of the related record
 * @example
 * ```typescript
 * // Single relation
 * { Project: [Builder.relation("project-123")] }
 *
 * // Multiple relations
 * { Projects: [Builder.relation("project-123"), Builder.relation("project-456")] }
 * ```
 */
export function relation(primaryKey: string): RelationReference {
	return { type: "primaryKey", value: primaryKey };
}

/**
 * Validates a date string is in YYYY-MM-DD format and represents a valid date.
 */
function validateDateString(dateString: string): void {
	if (!/^\d{4}-\d{2}-\d{2}$/.test(dateString)) {
		throw new Error(
			`Invalid date format: ${dateString}. Expected YYYY-MM-DD format.`,
		);
	}

	const date = new Date(dateString);
	if (Number.isNaN(date.getTime())) {
		throw new Error(`Invalid date: ${dateString}`);
	}
}

/**
 * Internal helper to create a date mention token from a DateValue object.
 */
function createDateToken(dateValue: DateValue): TextValue {
	return [["\u2023", [["d", dateValue]]]];
}

/**
 * Parses an ISO 8601 datetime string into date and time components.
 * Accepts formats like "2024-01-15T10:30", "2024-01-15T10:30:00Z",
 * "2024-01-15T10:30:00.000+05:00", etc.
 */
function parseISODateTime(isoString: string): { date: string; time: string } {
	if (!/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}/.test(isoString)) {
		throw new Error(
			`Invalid ISO 8601 datetime: ${isoString}. Expected format like 2024-01-15T10:30 or 2024-01-15T10:30:00Z`,
		);
	}

	const date = isoString.slice(0, 10);
	const time = isoString.slice(11, 16);

	if (Number.isNaN(new Date(isoString).getTime())) {
		throw new Error(`Invalid datetime: ${isoString}`);
	}

	return { date, time };
}

/**
 * Creates an emoji icon.
 * @param emoji - An emoji string (e.g., "🎯", "✨", "🚀")
 */
export function emojiIcon(emoji: string): Icon {
	return {
		type: "emoji",
		value: emoji,
	};
}

/**
 * Creates a Notion icon with a specific color.
 * @param icon - The name of the Notion icon (e.g., "checkmark", "pizza", "rocket")
 * @param color - The color variant (defaults to "gray")
 */
export function notionIcon(
	icon: NoticonName,
	color: NoticonColor = "gray",
): Icon {
	return {
		type: "notion",
		icon,
		color,
	};
}

/**
 * Creates an image icon from an external URL.
 * @param url - The URL of the image (e.g., "https://example.com/icon.png")
 */
export function imageIcon(url: string): Icon {
	return {
		type: "image",
		url,
	};
}
