// @notionhq/workers/src/schema.ts
import type {
	DateFormat,
	Icon,
	NumberFormat,
	SelectOption,
	StatusGroup,
} from "./types.js";

/**
 * Supported property types for sync schemas.
 */
export type PropertyType =
	| "title"
	| "rich_text"
	| "url"
	| "email"
	| "phone_number"
	| "checkbox"
	| "file"
	| "number"
	| "date"
	| "select"
	| "multi_select"
	| "status"
	| "people"
	| "place";

/**
 * Definition of a single property in a sync schema.
 */
export type PropertyDefinition = {
	type: PropertyType;
};

/**
 * A schema defining the structure a single property in a data source.
 */
export type PropertyConfiguration =
	| { type: "title" }
	| { type: "text" }
	| { type: "url" }
	| { type: "email" }
	| { type: "phone_number" }
	| { type: "checkbox" }
	| { type: "file" }
	| {
			type: "number";
			format?: NumberFormat;
	  }
	| {
			type: "date";
			date_format?: DateFormat;
	  }
	| {
			type: "select";
			options: SelectOption[];
	  }
	| {
			type: "multi_select";
			options: SelectOption[];
	  }
	| {
			type: "status";
			groups: StatusGroup[];
	  }
	| { type: "people" }
	| { type: "place" }
	| {
			type: "relation";
			/**
			 * The export name of the sync capability that defines the related collection.
			 * This must match the export name used when defining the related sync capability.
			 */
			relatedSyncKey: string;
			config: { twoWay: false } | { twoWay: true; relatedPropertyName: string };
	  };

export type Schema<PK extends string> = {
	/**
	 * The default name for the database when it is first created.
	 *
	 * Updating this after the database has been created will not change the
	 * name of the database.
	 */
	defaultName: string;
	/**
	 * Optional icon to use as the icon for the database page.
	 * If not provided, defaults to 📋.
	 * Use the `icon()` builder to create an icon value.
	 */
	databaseIcon?: Icon;
	properties: PropertySchema<PK>;
	/**
	 * Optional configuration for sub-item relations.
	 * If provided, the parent and child properties must be part of a two-way self-relation.
	 */
	subItems?: {
		parentPropertyName: string;
		childPropertyName: string;
	};
};

/**
 * A schema defining the structure of properties in a data source.
 */
export type PropertySchema<PK extends string> = {
	[PrimaryKey in PK]: PropertyConfiguration;
} & {
	[propertyName: string]: PropertyConfiguration;
};

/**
 * Creates a title property definition.
 */
export function title(): PropertyConfiguration {
	return { type: "title" };
}

/**
 * Creates a rich text property definition.
 */
export function richText(): PropertyConfiguration {
	return { type: "text" };
}

/**
 * Creates a URL property definition.
 */
export function url(): PropertyConfiguration {
	return { type: "url" };
}

/**
 * Creates an email property definition.
 */
export function email(): PropertyConfiguration {
	return { type: "email" };
}

/**
 * Creates a phone number property definition.
 */
export function phoneNumber(): PropertyConfiguration {
	return { type: "phone_number" };
}

/**
 * Creates a checkbox property definition.
 */
export function checkbox(): PropertyConfiguration {
	return { type: "checkbox" };
}

/**
 * Creates a file property definition.
 */
export function file(): PropertyConfiguration {
	return { type: "file" };
}

/**
 * Creates a number property definition with optional formatting.
 */
export function number(format?: NumberFormat): PropertyConfiguration {
	return format ? { type: "number", format } : { type: "number" };
}

/**
 * Creates a date property definition with optional formatting.
 */
export function date(date_format?: DateFormat): PropertyConfiguration {
	return date_format ? { type: "date", date_format } : { type: "date" };
}

/**
 * Creates a select property definition with predefined options.
 */
export function select(options: SelectOption[]): PropertyConfiguration {
	return { type: "select", options };
}

/**
 * Creates a multi-select property definition with predefined options.
 */
export function multiSelect(options: SelectOption[]): PropertyConfiguration {
	return { type: "multi_select", options };
}

/**
 * Creates a status property definition with groups.
 */
export function status(config: {
	groups: StatusGroup[];
}): PropertyConfiguration {
	return { type: "status", groups: config.groups };
}

/**
 * Creates a people property definition.
 */
export function people(): PropertyConfiguration {
	return { type: "people" };
}

/**
 * Creates a place property definition for storing geographic locations.
 */
export function place(): PropertyConfiguration {
	return { type: "place" };
}

/**
 * Creates a relation property definition that references another sync capability.
 * The related collection must be defined by a sync capability in the same worker.
 *
 * @param relatedSyncKey - The export name of the sync capability that defines the related collection.
 * @example
 * ```typescript
 * export const projectsSync = sync({...});
 *
 * export const tasksSync = sync({
 *   schema: {
 *     properties: {
 *       "Task Title": Schema.title(),
 *       "Project": Schema.relation("projectsSync"),
 *     },
 *   },
 *   ...
 * });
 * ```
 */
export function relation(
	relatedSyncKey: string,
	config?: { twoWay: false } | { twoWay: true; relatedPropertyName: string },
): PropertyConfiguration {
	return {
		type: "relation",
		relatedSyncKey,
		config: config ?? { twoWay: false },
	};
}
