// @notionhq/workers/src/types.ts
/**
 * A JSON value is a string, number, boolean, null, array, or object.
 */
export type JSONValue =
	| string
	| number
	| boolean
	| null
	| JSONValue[]
	| { [key: string]: JSONValue };

/**
 * A text token is a tuple where the first element is the text content
 * and optional subsequent elements are formatting annotations.
 */
type TextToken = [string, ...unknown[]];

/**
 * A TextValue is an array of text tokens, representing rich text content.
 */
export type TextValue = Array<TextToken>;

/**
 * Select option color types supported by Notion
 */
export type SelectColor =
	| "default"
	| "gray"
	| "brown"
	| "orange"
	| "yellow"
	| "green"
	| "blue"
	| "purple"
	| "pink"
	| "red";

/**
 * A select option with an optional color
 */
export interface SelectOption {
	name: string;
	color?: SelectColor;
}

/**
 * Status group types in Notion
 */
export type StatusGroupType = "To-do" | "In progress" | "Complete";

/**
 * Status group configuration
 */
export interface StatusGroup {
	name: StatusGroupType;
	options: SelectOption[];
}

/**
 * Date object representing a single date without time
 */
export interface NotionDate {
	type: "date";
	start_date: string; // Format: "YYYY-MM-DD"
}

/**
 * Date object representing a date with time
 */
export interface NotionDateTime {
	type: "datetime";
	start_date: string; // Format: "YYYY-MM-DD"
	start_time: string; // Format: "HH:mm"
	time_zone?: string; // e.g., "America/Los_Angeles"
}

/**
 * Date object representing a date range
 */
export interface DateRange {
	type: "daterange";
	start_date: string; // Format: "YYYY-MM-DD"
	end_date: string; // Format: "YYYY-MM-DD"
}

/**
 * Date object representing a date-time range
 */
export interface DateTimeRange {
	type: "datetimerange";
	start_date: string; // Format: "YYYY-MM-DD"
	start_time: string; // Format: "HH:mm"
	end_date: string; // Format: "YYYY-MM-DD"
	end_time: string; // Format: "HH:mm"
	time_zone?: string;
}

/**
 * All possible date value types
 */
export type DateValue = NotionDate | NotionDateTime | DateRange | DateTimeRange;

/**
 * Number format types
 */
export type NumberFormat =
	| "number"
	| "number_with_commas"
	| "percent"
	| "dollar"
	| "euro"
	| "pound"
	| "yen"
	| "rupee"
	| "won"
	| "yuan";

/**
 * Date format types
 */
export type DateFormat =
	| "relative"
	| "MM/DD/YYYY"
	| "DD/MM/YYYY"
	| "YYYY/MM/DD"
	| "ll"
	| "MMM d";

import type { NoticonName } from "./icon-names.js";

/**
 * Icon representing an emoji
 */
export interface EmojiIcon {
	type: "emoji";
	value: string;
}

/**
 * Valid colors for Notion icons
 */
export type NoticonColor =
	| "gray"
	| "lightgray"
	| "brown"
	| "yellow"
	| "orange"
	| "green"
	| "blue"
	| "purple"
	| "pink"
	| "red";

/**
 * Icon representing a Notion built-in icon
 */
export interface NoticonIcon {
	type: "notion";
	/**
	 * The name of the Notion icon (e.g., "checkmark", "pizza", "rocket")
	 */
	icon: NoticonName;
	/**
	 * The color variant of the icon
	 */
	color: NoticonColor;
}

/**
 * Icon representing an external image URL
 */
export interface ImageIcon {
	type: "image";
	/**
	 * The URL of the image (must be a valid http/https URL)
	 */
	url: string;
}

/**
 * All possible icon types
 */
export type Icon = EmojiIcon | NoticonIcon | ImageIcon;

/**
 * Person reference - represents a user in a people property
 */
export interface PersonReference {
	/**
	 * The email address of the user
	 */
	email: string;
}

/**
 * A PeopleValue is an array of person references
 */
export type PeopleValue = PersonReference[];

export type { NoticonName } from "./icon-names.js";

/**
 * Place value representing a geographic location.
 * Used for place properties in databases.
 */
export interface PlaceValue {
	/**
	 * Latitude coordinate (required)
	 */
	lat: number;
	/**
	 * Longitude coordinate (required)
	 */
	lon: number;
	/**
	 * Optional display name for the location
	 */
	name?: string;
	/**
	 * Optional street address
	 */
	address?: string;
	/**
	 * Optional Google Place ID for the location
	 */
	googlePlaceId?: string;
}

/**
 * A reference to a related record by its primary key.
 */
export type RelationReference = {
	type: "primaryKey";
	value: string;
};

/**
 * Relation value representing references to related records.
 * Each reference identifies a record in the target collection.
 */
export type RelationValue = RelationReference[];

/**
 * Time units for schedule intervals.
 * - "m": minutes
 * - "h": hours
 * - "d": days
 */
export type TimeUnit = "m" | "h" | "d";

/**
 * A string representing an interval, e.g. "30m", "1h", "7d".
 */
export type IntervalString = `${number}${TimeUnit}`;

/**
 * Schedule configuration for sync capabilities.
 * - "continuous": Run as frequently as the system allows
 * - IntervalString: Run at specified intervals, e.g. "30m", "1h", "1d"
 */
export type Schedule = "continuous" | IntervalString;

/**
 * Normalized schedule representation stored in the backend.
 */
export type SyncSchedule =
	| { type: "continuous" }
	| { type: "interval"; intervalMs: number };

export type HandlerOptions = {
	concreteOutput?: true;
};
