// @notionhq/workers/dist/builder.js
function richText(content) {
  return [[content]];
}
function url(url2) {
  return [[url2]];
}
function title(content) {
  return [[content]];
}
function text(content) {
  return [[content]];
}
function email(email2) {
  return [[email2]];
}
function phoneNumber(phone) {
  return [[phone]];
}
function checkbox(checked) {
  return checked ? [["Yes"]] : [["No"]];
}
function file(fileUrl, fileName) {
  return [[fileName ?? fileUrl, [["a", fileUrl]]]];
}
function number(value) {
  if (Number.isNaN(value)) {
    return [];
  }
  return [[value.toString()]];
}
function date(dateString) {
  validateDateString(dateString);
  const dateValue = {
    type: "date",
    start_date: dateString
  };
  return createDateToken(dateValue);
}
function dateTime(isoString, timeZone) {
  const { date: date2, time } = parseISODateTime(isoString);
  const dateValue = {
    type: "datetime",
    start_date: date2,
    start_time: time
  };
  if (timeZone) {
    dateValue.time_zone = timeZone;
  }
  return createDateToken(dateValue);
}
function dateRange(startDate, endDate) {
  validateDateString(startDate);
  validateDateString(endDate);
  const dateValue = {
    type: "daterange",
    start_date: startDate,
    end_date: endDate
  };
  return createDateToken(dateValue);
}
function dateTimeRange(startDateTime, endDateTime, timeZone) {
  const start = parseISODateTime(startDateTime);
  const end = parseISODateTime(endDateTime);
  const dateValue = {
    type: "datetimerange",
    start_date: start.date,
    start_time: start.time,
    end_date: end.date,
    end_time: end.time
  };
  if (timeZone) {
    dateValue.time_zone = timeZone;
  }
  return createDateToken(dateValue);
}
function link(displayText, url2) {
  return [[displayText, [["a", url2]]]];
}
function select(value) {
  return [[value]];
}
function multiSelect(...values) {
  if (values.length === 0) {
    return [];
  }
  return [[values.join(",")]];
}
function status(value) {
  return [[value]];
}
function people(...emails) {
  return emails.map((email2) => ({ email: email2 }));
}
function place(value) {
  if (typeof value.lat !== "number" || typeof value.lon !== "number") {
    throw new Error("Place value must have numeric lat and lon coordinates");
  }
  return value;
}
function relation(primaryKey) {
  return { type: "primaryKey", value: primaryKey };
}
function validateDateString(dateString) {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(dateString)) {
    throw new Error(
      `Invalid date format: ${dateString}. Expected YYYY-MM-DD format.`
    );
  }
  const date2 = new Date(dateString);
  if (Number.isNaN(date2.getTime())) {
    throw new Error(`Invalid date: ${dateString}`);
  }
}
function createDateToken(dateValue) {
  return [["\u2023", [["d", dateValue]]]];
}
function parseISODateTime(isoString) {
  if (!/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}/.test(isoString)) {
    throw new Error(
      `Invalid ISO 8601 datetime: ${isoString}. Expected format like 2024-01-15T10:30 or 2024-01-15T10:30:00Z`
    );
  }
  const date2 = isoString.slice(0, 10);
  const time = isoString.slice(11, 16);
  if (Number.isNaN(new Date(isoString).getTime())) {
    throw new Error(`Invalid datetime: ${isoString}`);
  }
  return { date: date2, time };
}
function emojiIcon(emoji) {
  return {
    type: "emoji",
    value: emoji
  };
}
function notionIcon(icon, color = "gray") {
  return {
    type: "notion",
    icon,
    color
  };
}
function imageIcon(url2) {
  return {
    type: "image",
    url: url2
  };
}
export {
  checkbox,
  date,
  dateRange,
  dateTime,
  dateTimeRange,
  email,
  emojiIcon,
  file,
  imageIcon,
  link,
  multiSelect,
  notionIcon,
  number,
  people,
  phoneNumber,
  place,
  relation,
  richText,
  select,
  status,
  text,
  title,
  url
};
