// @notionhq/workers/dist/schema.js
function title() {
  return { type: "title" };
}
function richText() {
  return { type: "text" };
}
function url() {
  return { type: "url" };
}
function email() {
  return { type: "email" };
}
function phoneNumber() {
  return { type: "phone_number" };
}
function checkbox() {
  return { type: "checkbox" };
}
function file() {
  return { type: "file" };
}
function number(format) {
  return format ? { type: "number", format } : { type: "number" };
}
function date(date_format) {
  return date_format ? { type: "date", date_format } : { type: "date" };
}
function select(options) {
  return { type: "select", options };
}
function multiSelect(options) {
  return { type: "multi_select", options };
}
function status(config) {
  return { type: "status", groups: config.groups };
}
function people() {
  return { type: "people" };
}
function place() {
  return { type: "place" };
}
function relation(relatedSyncKey, config) {
  return {
    type: "relation",
    relatedSyncKey,
    config: config ?? { twoWay: false }
  };
}
export {
  checkbox,
  date,
  email,
  file,
  multiSelect,
  number,
  people,
  phoneNumber,
  place,
  relation,
  richText,
  select,
  status,
  title,
  url
};
