// @notionhq/workers/dist/worker.js
import { createAutomationCapability } from "./capabilities/automation.js";
import { createOAuthCapability } from "./capabilities/oauth.js";
import { createSyncCapability } from "./capabilities/sync.js";
import { createToolCapability } from "./capabilities/tool.js";
class Worker {
  #capabilities = /* @__PURE__ */ new Map();
  /**
   * Register a sync capability.
   *
   * Example:
   *
   * ```ts
   * import { Worker } from "@notionhq/workers";
   * import * as Builder from "@notionhq/workers/builder";
   * import * as Schema from "@notionhq/workers/schema";
   *
   * const worker = new Worker();
   * export default worker;
   *
   * worker.sync("tasksSync", {
   *   primaryKeyProperty: "Task ID",
   *   schema: {
   *     defaultName: "Tasks",
   *     properties: {
   *       "Task Name": Schema.title(),
   *       "Task ID": Schema.richText(),
   *       Status: Schema.select([
   *         { name: "Open", color: "default" },
   *         { name: "Done", color: "green" },
   *       ]),
   *     },
   *   },
   *   execute: async () => {
   *     const changes = [
   *       {
   *         key: "task-1",
   *         properties: {
   *           "Task Name": Builder.title("Write docs"),
   *           "Task ID": Builder.richText("task-1"),
   *           Status: Builder.select("Open"),
   *         },
   *       },
   *     ];
   *
   *     return { changes, hasMore: false };
   *   },
   * });
   * ```
   *
   * @param key - The unique key for this capability.
   * @param config - The sync configuration.
   * @returns The capability object.
   */
  sync(key, config) {
    this.#validateUniqueKey(key);
    const capability = createSyncCapability(key, config);
    this.#capabilities.set(key, capability);
    return capability;
  }
  /**
   * Register a tool capability.
   *
   * Example:
   *
   * ```ts
   * import { j } from "@notionhq/workers/schema-builder";
   *
   * worker.tool<{ name: string }, string>("sayHello", {
   *   title: "Say Hello",
   *   description: "Say hello to the user",
   *   schema: j.object({ name: j.string() }),
   *   execute: ({ name }, { notion }) => {
   *     return `Hello, ${name}!`;
   *   },
   * })
   * ```
   *
   * @param key - The unique key for this capability.
   * @param config - The tool configuration.
   * @returns The capability object.
   */
  tool(key, config) {
    this.#validateUniqueKey(key);
    const capability = createToolCapability(key, config);
    this.#capabilities.set(key, capability);
    return capability;
  }
  /**
   * Register an automation capability.
   *
   * Example:
   *
   * ```ts
   * const worker = new Worker();
   * export default worker;
   *
   * worker.automation("sendWelcomeEmail", {
   *   title: "Send Welcome Email",
   *   description: "Sends a welcome email when a new user is added",
   *   execute: async (event, { notion }) => {
   *     const { pageId, pageData } = event;
   *
   *     // Access page properties from the Public API format
   *     if (pageData) {
   *       const name = pageData.properties.Name; // Access any property
   *       const status = pageData.properties.Status;
   *       console.log(`Processing: ${name}`);
   *     }
   *
   *     // Your automation logic here
   *     await sendEmail(pageId);
   *   },
   * })
   * ```
   *
   * @param key - The unique key for this capability.
   * @param config - The automation configuration.
   * @returns The capability object.
   */
  automation(key, config) {
    this.#validateUniqueKey(key);
    const capability = createAutomationCapability(key, config);
    this.#capabilities.set(key, capability);
    return capability;
  }
  /**
   * Register an OAuth capability.
   *
   * There are two ways to configure OAuth:
   *
   * 1. Notion-managed providers:
   * ```ts
   * const worker = new Worker();
   * export default worker;
   *
   * worker.oauth("googleAuth", {
   *   type: "notion_managed",
   *   name: "my-google-auth",
   *   provider: "google"
   * })
   * ```
   *
   * 2. User-managed OAuth configuration:
   * ```ts
   * const worker = new Worker();
   * export default worker;
   *
   * worker.oauth("myCustomAuth", {
   *   type: "user_managed",
   *   name: "my-custom-oauth",
   *   authorizationEndpoint: "https://provider.com/oauth/authorize",
   *   tokenEndpoint: "https://provider.com/oauth/token",
   *   scope: "read write",
   *   clientId: process.env.CLIENT_ID,
   *   clientSecret: process.env.CLIENT_SECRET,
   *   authorizationParams: {
   *     access_type: "offline",
   *     prompt: "consent"
   *   }
   * })
   * ```
   *
   * @param key - The unique key used to register this OAuth capability.
   * @param config - The OAuth configuration (Notion-managed or user-managed) for this capability.
   * @returns The registered OAuth capability.
   */
  oauth(key, config) {
    this.#validateUniqueKey(key);
    const capability = createOAuthCapability(key, config);
    this.#capabilities.set(key, capability);
    return capability;
  }
  /**
   * Get all registered capabilities (for discovery) without their handlers.
   */
  get capabilities() {
    return Array.from(this.#capabilities.values()).map((c) => ({
      _tag: c._tag,
      key: c.key,
      config: c.config
    }));
  }
  /**
   * Execute a capability by key.
   *
   * @param key - The key of the capability to execute.
   * @param context - The context to pass to the capability.
   * @param options - Additional options for execution (e.g. for testing).
   * @returns The result of the capability execution.
   */
  async run(key, context, options = {}) {
    const capability = this.#capabilities.get(key);
    if (!capability) {
      throw new Error(`Capability "${key}" not found`);
    }
    if (capability._tag === "oauth") {
      throw new Error(
        `Cannot run OAuth capability "${key}" - OAuth capabilities only provide configuration`
      );
    }
    return capability.handler(context, options);
  }
  #validateUniqueKey(key) {
    if (!key || typeof key !== "string") {
      throw new Error("Capability key must be a non-empty string");
    }
    if (this.#capabilities.has(key)) {
      throw new Error(`Capability with key "${key}" already registered`);
    }
  }
}
export {
  Worker
};
